presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java

```diff
@@ -383,26 +390,56 @@ private synchronized Optional<QueryResults> getCachedResult(long token)
     private synchronized QueryResults getNextResultWithRetry(long token, UriInfo uriInfo, String scheme, DataSize targetResultSize, boolean binaryResults)
     {
         QueryResults queryResults = getNextResult(token, uriInfo, scheme, targetResultSize, binaryResults);
-        if (queryResults.getError() == null || !queryResults.getError().isRetriable()) {
-            return queryResults;
+
+        boolean historyBasedOptimizationEnabled = useHistoryBasedPlanStatisticsEnabled(session) && trackHistoryBasedPlanStatisticsEnabled(session);
+        boolean underOneRetryCount = queryManager.getQueryRetryCount(queryId) < 1;
+        boolean underRetryExecutionTime = queryManager.getQueryInfo(queryId).getQueryStats().getExecutionTime().toMillis() <= getQueryRetryMaxExecutionTime(session).toMillis();
+
+        if (queryResults.getError() != null && historyBasedOptimizationEnabled && underOneRetryCount && underRetryExecutionTime && retryQueryWithHistoryBasedOptimizationEnabled(session)) {
+            CanonicalPlanWithInfo currentQueryTopLevelPlanHash = queryManager.getFullQueryInfo(queryId).getPlanCanonicalInfo().get(0);
+
+            if (originalBeforeRetryQueryId == null && previousQueryTopLevelPlanHash == null) {
+                originalBeforeRetryQueryId = queryId;
+                previousQueryTopLevelPlanHash = currentQueryTopLevelPlanHash;
+            }
         }
+        else if (queryManager.getQueryRetryCount(queryId) == 1 && retryQueryWithHistoryBasedOptimizationEnabled(session)) {
+            CanonicalPlanWithInfo currentQueryTopLevelPlanHash = queryManager.getFullQueryInfo(queryId).getPlanCanonicalInfo().get(0);
```

Comment:
Directly use the existing hash may not work, at least for changes made in DeterminJoinDistributionType, as the hash is taken before this broadcast

Reply:
@feilong-liu Will add this as an issue and work on it in a separate PR as discussed